examples/sst2/models.py

```diff
@@ -0,0 +1,385 @@
+    self.dropout_layer = nn.Dropout(
+        rate=self.dropout_rate,
+        deterministic=self.deterministic)
+
+  def __call__(self, inputs: Array):
+    """Applies the MLP to the last dimension of the inputs.
+
+    Args:
+      inputs: <float32>[batch_size, ..., input_features].
+
+    Returns:
+      The MLP output <float32>[batch_size, ..., output_size]
+    """
+    hidden = self.intermediate_layer(inputs)
+    hidden = self.activation(hidden)
+    hidden = self.dropout_layer(hidden)
+    output = self.output_layer(hidden)
+    return output
+
+
+class KeysOnlyMlpAttention(nn.Module):
+  """Computes MLP-based attention scores based on keys alone, without a query.
+
+  Attention scores are computed by feeding the keys through an MLP. This
+  results in a single scalar per key, and for each sequence the attention
+  scores are normalized using a softmax so that they sum to 1. Invalid key
+  positions are ignored as indicated by the mask. This is also called
+  "Bahdanau attention" and was originally proposed in:
+  ```
+  Bahdanau et al., 2015. Neural Machine Translation by Jointly Learning to
+  Align and Translate. ICLR. https://arxiv.org/abs/1409.0473
+  ```
+
+  Attributes:
+    hidden_size: The hidden size of the MLP that computes the attention score.
+  """
+  hidden_size: int
+
+  @nn.compact
+  def __call__(self, keys: Array, mask: Array) -> Array:
+    """Applies model  to the input keys and mask.
+
+    Args:
+      keys: The inputs for which to compute an attention score. Shape:
+        <float32>[batch_size, seq_length, embeddings_size].
+      mask: A mask that determinines which values in `keys` are valid. Only
+        values for which the mask is True will get non-zero attention scores.
+        <bool>[batch_size, seq_length].
+
+    Returns:
+      The normalized attention scores. <float32>[batch_size, seq_length].
+    """
+    hidden = nn.Dense(self.hidden_size, name='keys', use_bias=False)(keys)
+    energy = nn.tanh(hidden)
+    scores = nn.Dense(1, name='energy', use_bias=False)(energy)
+    scores = scores.squeeze(-1)  # New shape: <float32>[batch_size, seq_len].
+    scores = jnp.where(mask, scores, -jnp.inf)  # Using exp(-inf) = 0 below.
+    scores = nn.softmax(scores, axis=-1)
+
+    # Captures the scores if 'intermediates' is mutable, otherwise does nothing.
+    self.sow('intermediates', 'attention', scores)
+
+    return scores
+
+
+class AttentionClassifier(nn.Module):
+  """A classifier that uses attention to summarize the inputs.
+
+  Attributes:
+    hidden_size: The hidden size of the MLP classifier.
+    output_size: The number of output classes for the classifier.
+    dropout_rate: The dropout rate applied over the encoded_inputs, the summary
+      of the inputs, and inside the MLP. Applied when `deterministic` is False.
+    deterministic: Disables dropout if True.
+  """
+  hidden_size: int
+  output_size: int
+  dropout_rate: float = 0.
+  deterministic: bool = False
+
+  def setup(self):
```

Comment:
Out of curiosity: Is there a reason you are using `setup` rather than `@nn.compact` here? I guess for readability? (Same for `TextClassifier` below)

Reply:
Yes, especially for TextClassifier it allows calling the various sub-methods such as `embed_token_ids` and `logits_from_embedded_inputs`. They'll enable some cool use cases in the Colab :)